File: crates/codeprysm-core/tests/fixtures/integration_repos/csharp/sample.cs
/**
 * Sample C# module for integration testing.
 *
 * This module demonstrates various C# language features for
 * graph generation validation including classes, interfaces, and generics.
 */

using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace IntegrationTests
{
    // Module-level constant
    public static class Constants
    {
        public const int MaxItems = 100;
    }

    // Enum definition
    public enum UserRole
    {
        Admin,
        User,
        Guest
    }

    // Interface definition
    public interface ICalculator
    {
        int Value { get; }
        int Add(int amount);
        int Multiply(int factor);
    }

    // Generic interface
    public interface IRepository<T> where T : class
    {
        Task<T?> FindById(string id);
        Task<IEnumerable<T>> FindAll();
        Task<T> Save(T item);
        Task<bool> Delete(string id);
    }

    // Class implementing interface
    public class Calculator : ICalculator
    {
        private int _value;
        private readonly List<int> _history;

        public Calculator(int initialValue = 0)
        {
            _value = initialValue;
            _history = new List<int>();
        }

        public int Value => _value;

        public IReadOnlyList<int> History => _history.AsReadOnly();

        public int Add(int amount)
        {
            _value += amount;
            _history.Add(amount);
            return _value;
        }

        public int Multiply(int factor)
        {
            _value *= factor;
            return _value;
        }

        public static int Square(int x) => x * x;
    }

    // Class with async methods
    public class AsyncProcessor
    {
        private readonly string _name;
        private int _processedCount;

        public AsyncProcessor(string name)
        {
            _name = name;
            _processedCount = 0;
        }

        public string Name => _name;
        public int ProcessedCount => _processedCount;

        public async Task<string> ProcessItem(string item)
        {
            await Task.Delay(10);
            _processedCount++;
            return $"{_name}:{item}";
        }

        public async Task<List<string>> ProcessBatch(IEnumerable<string> items)
        {
            var results = new List<string>();
            foreach (var item in items)
            {
                var result = await ProcessItem(item);
                results.Add(result);
            }
            return results;
        }
    }

    // Generic class
    public class DataProcessor<T>
    {
        private readonly List<T> _data;

        public DataProcessor()
        {
            _data = new List<T>();
        }

        public void Add(T item)
        {
            _data.Add(item);
        }

        public List<TResult> Map<TResult>(Func<T, TResult> selector)
        {
            var results = new List<TResult>();
            foreach (var item in _data)
            {
                results.Add(selector(item));
            }
            return results;
        }

        public List<T> Filter(Func<T, bool> predicate)
        {
            var results = new List<T>();
            foreach (var item in _data)
            {
                if (predicate(item))
                {
                    results.Add(item);
                }
            }
            return results;
        }
    }

    // Abstract base class
    public abstract class BaseService
    {
        protected string Name { get; }

        protected BaseService(string name)
        {
            Name = name;
        }

        public abstract Task Initialize();

        public virtual string GetName() => Name;
    }

    // Concrete implementation
    public class UserService : BaseService
    {
        private readonly Dictionary<string, User> _users;

        public UserService() : base("UserService")
        {
            _users = new Dictionary<string, User>();
        }

        public override async Task Initialize()
        {
            await Task.CompletedTask;
        }

        public async Task<User> CreateUser(string name, string email)
        {
            var user = new User
            {
                Id = Guid.NewGuid().ToString(),
                Name = name,
                Email = email,
                CreatedAt = DateTime.UtcNow
            };
            _users[user.Id] = user;
            return await Task.FromResult(user);
        }
    }

    // Record type (C# 9+)
    public record User
    {
        public string Id { get; init; } = "";
        public string Name { get; init; } = "";
        public string Email { get; init; } = "";
        public DateTime CreatedAt { get; init; }
    }

    // Static class with extension methods
    public static class StringExtensions
    {
        public static string Reverse(this string input)
        {
            var chars = input.ToCharArray();
            Array.Reverse(chars);
            return new string(chars);
        }
    }

    // Standalone function (as static method)
    public static class Functions
    {
        public static int StandaloneFunction(string param)
        {
            return param?.Length ?? 0;
        }

        public static async Task<Dictionary<string, object>> AsyncStandalone(string url)
        {
            await Task.Delay(100);
            return new Dictionary<string, object> { { "url", url } };
        }
    }
}

File: crates/codeprysm-core/tests/fixtures/integration_repos/typescript/sample.ts
/**
 * Sample TypeScript module for integration testing.
 *
 * This module demonstrates various TypeScript language features for
 * graph generation validation including interfaces, types, and generics.
 */

// Type alias
type UserId = string;

// Generic type alias
type Result<T, E = Error> = { ok: true; value: T } | { ok: false; error: E };

// Interface definition
interface User {
    id: UserId;
    name: string;
    email: string;
    createdAt: Date;
}

// Interface with methods
interface Repository<T> {
    findById(id: string): Promise<T | null>;
    findAll(): Promise<T[]>;
    save(item: T): Promise<T>;
    delete(id: string): Promise<boolean>;
}

// Enum definition
enum UserRole {
    Admin = "admin",
    User = "user",
    Guest = "guest"
}

// Module-level constant
const MAX_ITEMS: number = 100;

/**
 * A standalone function outside any class.
 */
function standaloneFunction(param: string): string {
    return `processed_${param}`;
}

/**
 * An async standalone function.
 */
async function asyncStandalone(url: string): Promise<{ url: string }> {
    await new Promise(resolve => setTimeout(resolve, 100));
    return { url };
}

/**
 * Generic function
 */
function identity<T>(value: T): T {
    return value;
}

/**
 * Arrow function with type annotations.
 */
const arrowFunction = (x: number, y: number): number => x + y;

/**
 * A generic class implementing Repository interface.
 */
class UserRepository implements Repository<User> {
    private items: Map<string, User>;

    constructor() {
        this.items = new Map();
    }

    async findById(id: string): Promise<User | null> {
        return this.items.get(id) ?? null;
    }

    async findAll(): Promise<User[]> {
        return Array.from(this.items.values());
    }

    async save(user: User): Promise<User> {
        this.items.set(user.id, user);
        return user;
    }

    async delete(id: string): Promise<boolean> {
        return this.items.delete(id);
    }
}

/**
 * A class with generic methods.
 */
class DataProcessor<T> {
    private data: T[];

    constructor(initialData: T[] = []) {
        this.data = initialData;
    }

    add(item: T): void {
        this.data.push(item);
    }

    map<U>(fn: (item: T) => U): U[] {
        return this.data.map(fn);
    }

    filter(predicate: (item: T) => boolean): T[] {
        return this.data.filter(predicate);
    }
}

/**
 * Abstract base class.
 */
abstract class BaseService {
    protected readonly name: string;

    constructor(name: string) {
        this.name = name;
    }

    abstract initialize(): Promise<void>;

    getName(): string {
        return this.name;
    }
}

/**
 * Concrete implementation of abstract class.
 */
class UserService extends BaseService {
    private userRepo: UserRepository;

    constructor() {
        super("UserService");
        this.userRepo = new UserRepository();
    }

    async initialize(): Promise<void> {
        // Initialize service
    }

    async createUser(name: string, email: string): Promise<User> {
        const user: User = {
            id: crypto.randomUUID(),
            name,
            email,
            createdAt: new Date()
        };
        return this.userRepo.save(user);
    }
}

// Export all types, interfaces, and classes
export {
    UserId,
    Result,
    User,
    Repository,
    UserRole,
    MAX_ITEMS,
    standaloneFunction,
    asyncStandalone,
    identity,
    arrowFunction,
    UserRepository,
    DataProcessor,
    BaseService,
    UserService
};
